### PR TITLE
Control e2e framework -> k/k/pkg dependencies

### PR DESCRIPTION
**Pull request by bart0sh**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

- Got rid of the last dependency to k/k/pkg/util
- Enabled control of the k/k/pkg imports

#### Special notes for your reviewer:

After this cleanup there are just a few exceptions that we need to fix, like k8s.io/kubernetes/pkg/api/v1/pod for some submodules.
Another good thing is that all k/k/pkg imports are now under control. Adding new ones would require updating .import-restrictions.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/triage accepted
/priority important-longterm
/assign @pohly @dims

### CODE REVIEW

**Comment by aojea** [comment]
the verify error seems legit

**Comment by pohly** [comment]
We can move forward by granting exceptions for those sub-packages that still access to k/k.

Then later someone can clean those up or we just leave it. However, at least using k8s.io/kubernetes/pkg/features in test/e2e/framework/pv is suspicious. I wouldn't be surprised if some code in the E2E suite incorrectly checked feature gates, directly or indirectly...

**Comment by aojea** [comment]
+1

**Comment by bart0sh** [comment]
> We can move forward by granting exceptions for those sub-packages that still access to k/k.
> 
> Then later someone can clean those up or we just leave it. However, at least using k8s.io/kubernetes/pkg/features in test/e2e/framework/pv is suspicious. I wouldn't be surprised if some code in the E2E suite incorrectly checked feature gates, directly or indirectly...

There are a lot of suspicious deps there. Could it be that something is wrong with the checker? Most of missed dependencies reported by the pull-kubernetes-verify are not direct e2e framework dependencies, for example `k8s.io/kubernetes/pkg/securitycontext` are not actually imported by the framework code:

```
$ git grep securitycontext ./test/e2e/framework/
test/e2e/framework/daemonset/.import-restrictions:    - k8s.io/kubernetes/pkg/securitycontext
test/e2e/framework/providers/gce/.import-restrictions:    - k8s.io/kubernetes/pkg/securitycontext
test/e2e/framework/providers/kubemark/.import-restrictions:    - k8s.io/kubernetes/pkg/securitycontext
test/e2e/framework/pv/.import-restrictions:    - k8s.io/kubernetes/pkg/securitycontext
```

P.S. Those additional dependencies appear in the output of ./hack/verify-import-boss.sh because `k8s.io/kubernetes/pkg/...` is used there. When I remove it from the script, [these dependencies](https://github.com/kubernetes/kubernetes/compare/ed98711d8d75571ebf3071d2f3107bf450fe6757..8706b4372efb8ae4ef6d8a79230643b838787151) are not reported.

**Comment by pohly** [comment]
import-boss also complains about indirect dependencies, and that's correct.

For example, test/e2e/framework/pv -> test/e2e/storage/utils -> pkg/volume/util -> pkg/volume/util/resize_util.go -> pkg/features.

**Comment by bart0sh** [comment]
> import-boss also complains about indirect dependencies, and that's correct.

I think it's confusing as if it shows only direct dependencies we'd know what to remove. With indirect dependencies .import-restrictions files are full of imports that don't present in the source code and therefore we can't do anything about those.

However, if you thinks it's correct then please merge this PR.

**Comment by pohly** [comment]
> I think it's confusing as if it shows only direct dependencies we'd know what to remove. 

It would be nicer if it showed the dependency graph, but that then begs the question what it should show if, e.g., package A depends on B and C, which both depend on D, where only D is a forbidden dependency of A.

Removing B and C may not be feasible. Both of those may have to be modified to not depend on D.

All of that is something that the developer needs to figure out.

File: test/e2e/framework/auth/.import-restrictions [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
-# dependencies, therefore we need to override the restrictions from
-# the parent .import-restrictions file.
+# dependencies except of k/k/pkg, therefore we need to override the
+# restrictions from the parent .import-restrictions file.
 #
 # At some point it may become useful to also check this package's
 # dependencies more careful.
 rules:
+  - selectorRegexp: "^k8s[.]io/kubernetes/pkg"
+    forbiddenPrefixes: [ "" ]
```
Comment (pohly) [inline test/e2e/framework/auth/.import-restrictions:9]: Instead of `forbiddenPrefixes: [ "" ]`, can you use `allowedPrefixes: []`? I find that clearer because it matches how one would describe this rule in plain English: "from k8s.io/kubernetes/pkg, allow nothing".

File: test/e2e/framework/providers/gce/.import-restrictions [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# This E2E framework sub-package is currently allowed to use arbitrary
+# dependencies except of k/k/pkg, therefore we need to override the
+# restrictions from the parent .import-restrictions file.
+#
+# At some point it may become useful to also check this package's
+# dependencies more careful.
+rules:
+  - selectorRegexp: "^k8s[.]io/kubernetes/pkg"
+    allowedPrefixes:
+    - k8s.io/kubernetes/pkg/api/legacyscheme
```
Comment (pohly) [inline test/e2e/framework/providers/gce/.import-restrictions:10]: Maintaining these lists will be problematic. Let's start doing that once we know that we want to limit usage of k/k/pkg in this package, i.e. for now don't add this file.

File: test/e2e/framework/providers/kubemark/.import-restrictions [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    - k8s.io/kubernetes/pkg/fieldpath
+    - k8s.io/kubernetes/pkg/kubelet
+    - k8s.io/kubernetes/pkg/probe
+    - k8s.io/kubernetes/pkg/proxy
+    - k8s.io/kubernetes/pkg/registry/core/service/allocator
+    - k8s.io/kubernetes/pkg/scheduler
+    - k8s.io/kubernetes/pkg/security/apparmor
+    - k8s.io/kubernetes/pkg/securitycontext
+    - k8s.io/kubernetes/pkg/util/
+    - k8s.io/kubernetes/pkg/volume
```
Comment (pohly) [inline test/e2e/framework/providers/kubemark/.import-restrictions:28]: Same here.

File: test/e2e/framework/pv/.import-restrictions [hunk 4]
```diff
@@ -1,9 +1,23 @@
...
+    - k8s.io/kubernetes/pkg/api/v1/pod
+    - k8s.io/kubernetes/pkg/api/legacyscheme
+    - k8s.io/kubernetes/pkg/apis/
+    - k8s.io/kubernetes/pkg/features
+    - k8s.io/kubernetes/pkg/kubelet/server/metrics
+    - k8s.io/kubernetes/pkg/proxy/util
+    - k8s.io/kubernetes/pkg/securitycontext
+    - k8s.io/kubernetes/pkg/util/parsers
+    - k8s.io/kubernetes/pkg/util/slice
+    - k8s.io/kubernetes/pkg/volume
```
Comment (pohly) [inline test/e2e/framework/pv/.import-restrictions:19]: Let's not choose specific allowed packages for now, i.e. leave this file as it was before.

File: test/e2e/framework/rc/.import-restrictions [hunk 5]
```diff
@@ -1,9 +1,13 @@
...
-# the parent .import-restrictions file.
+# dependencies except of k/k/pkg, therefore we need to override the
+# restrictions from the parent .import-restrictions file.
 #
 # At some point it may become useful to also check this package's
 # dependencies more careful.
 rules:
+  - selectorRegexp: "^k8s[.]io/kubernetes/pkg"
+    allowedPrefixes:
+    - k8s.io/kubernetes/pkg/api/v1/pod
```
Comment (pohly) [inline test/e2e/framework/rc/.import-restrictions:10]: Same here.

Comment (pohly) [reply test/e2e/framework/rc/.import-restrictions:10]: ... and below.

File: test/e2e/framework/pv/.import-restrictions (hunk 4 above)
Comment (bart0sh) [reply test/e2e/framework/pv/.import-restrictions:19]: There is only one direct dependency here: k8s.io/kubernetes/pkg/api/v1/pod, the rest is indirect.
Would it make sense to add a comment to separate indirect deps instead of allowing to add anything?

File: test/e2e/framework/rc/.import-restrictions (hunk 5 above)
Comment (bart0sh) [reply test/e2e/framework/rc/.import-restrictions:10]: I'd suggest to leave k8s.io/kubernetes/pkg/api/v1/pod as this the only dependency of this submodule. If we remove it we'll basically allow to add more deps. The overall idea of this PR was to add such a control.

File: test/e2e/framework/providers/gce/.import-restrictions (hunk 2 above)
Comment (bart0sh) [reply test/e2e/framework/providers/gce/.import-restrictions:10]: I agree about this one, but would leave k8s.io/kubernetes/pkg/api/v1/pod if it's the only one dependency.

File: test/e2e/framework/providers/kubemark/.import-restrictions (hunk 3 above)
Comment (bart0sh) [reply test/e2e/framework/providers/kubemark/.import-restrictions]: done

File: test/e2e/framework/pv/.import-restrictions (hunk 4 above)
Comment (bart0sh) [reply test/e2e/framework/pv/.import-restrictions]: done

File: test/e2e/framework/providers/gce/.import-restrictions (hunk 2 above)
Comment (bart0sh) [reply test/e2e/framework/providers/gce/.import-restrictions]: done

File: test/e2e/framework/auth/.import-restrictions (hunk 1 above)
Comment (bart0sh) [reply test/e2e/framework/auth/.import-restrictions]: done

**Comment by bart0sh** [comment]
@pohly only fully forbidden imports and `k8s.io/kubernetes/pkg/api/v1/pod` were left. These shouldn't be hard to maintain, I believe. PTAL.

File: test/e2e/framework/auth/.import-restrictions (hunk 1 above)
Comment (bart0sh) [reply test/e2e/framework/auth/.import-restrictions]: However, for me "forbid everything" sounds closer to what we want to do because it sounds stricter.

File: test/e2e/framework/rc/.import-restrictions (hunk 5 above)
Comment (pohly) [reply test/e2e/framework/rc/.import-restrictions]: Okay.

**Review by pohly — APPROVED** [review]
/lgtm
/approve

I wonder what it is that gets pulled from k8s.io/kubernetes/pkg/api/v1/pod. It might be something that would also be useful in a staging repo...

**Comment by bart0sh** [comment]
@pohly 
> I wonder what it is that gets pulled from k8s.io/kubernetes/pkg/api/v1/pod. It might be something that would also be useful in a staging repo...

All of them look quite useful from the first look. Here is the full list:
- IsPodAvailable returns true if a pod is available; false otherwise.
- IsPodReady returns true if a pod is ready; false otherwise.
- VisitContainers invokes the visitor function with a pointer to every container spec in the given pod spec with type set in mask. If visitor returns false, visiting is short-circuited. VisitContainers returns true if visiting completes, false if visiting was short-circuited.
- AllFeatureEnabledContainers returns a ContainerType mask which includes all container types except for the ones guarded by feature gate.

**Comment by aojea** [comment]
> k8s.io/kubernetes/pkg/api/v1/pod.

if it uses the pod API it should be using the external open `k8s.io/api/v1/pod`

There are also a lot of pod helpers in that  k8s.io/kubernetes/pkg/api/v1/pod folder, I would not be surprised is importing some of those, but helpers can be copied if necessary

**Comment by pohly** [comment]
Fully agreed. I created https://github.com/kubernetes/kubernetes/issues/117583 to track those next steps.
